### PR TITLE
Provides a cat api endpoint for templates.

### LINKED ISSUE #20467: Provide an easier way to list templates
Currently to get a list of the templates you have to call `_templates`, which then returns (usually) a large JSON result you need to parse.
This isn't the easiest way to find out what exists when you want to do a quick check or help someone troubleshoot.

Would be possible to add a `_cat/template`  output that listed something like name, index pattern (ie the `template value`) and the order so that users can quickly see what is in the cluster.
Otherwise a `_template?summary` option that shows the same?

### PR DESCRIPTION
**Pull request by a2lin**
Adds a cat api endpoint: `/_cat/templates` and its more specific version, `/_cat/templates/{name}`.

It looks something like:

```
$ curl "localhost:9200/_cat/templates?v"
name                  template     order version
sushi_california_roll *avocado*    1     1
pizza_hawaiian        *pineapples* 1
pizza_pepperoni       *pepperoni*  1
```

The specified version (only allows \* globs) looks like:

```
$ curl "localhost:9200/_cat/templates/pizza*"
name            template     order version
pizza_hawaiian  *pineapples* 1
pizza_pepperoni *pepperoni*  1
```

Partially specified columns:

```
$ curl "localhost:9200/_cat/templates/pizza*?v=true&h=name,template"
name            template
pizza_hawaiian  *pineapples*
pizza_pepperoni *pepperoni*
```

The help text:

```
$ curl "localhost:9200/_cat/templates/pizza*?help"
name     | n | template name
template | t | template pattern string
order    | o | template application order number
version  | v | version
```

Closes #20467

### CODE REVIEW

**Comment by clintongormley** [comment]
@tlrx could you review please?

**Comment by s1monw** [comment]
LGTM

**Comment by s1monw** [comment]
test this please

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTemplatesAction.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTemplatesAction.java]: I think it can fit on one line

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTemplatesAction.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTemplatesAction.java]: I think it could be null instead of `"*"`

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTemplatesAction.java
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTemplatesAction.java]: And here, we check `Regex.simpleMatch()` only if the pattern if not null?

File: docs/reference/cat/templates.asciidoc
Comment (tlrx) [inline docs/reference/cat/templates.asciidoc]: Maybe changed to "The `templates` command provides about existing templates" ?

File: docs/reference/cat/templates.asciidoc
Comment (tlrx) [inline docs/reference/cat/templates.asciidoc]: three templates active -> three existing templates ? Templates cannot be active or deactivated

File: docs/reference/indices/templates.asciidoc
Comment (tlrx) [inline docs/reference/indices/templates.asciidoc]: I think you can remove this because it's going to be documented in #20407

File: rest-api-spec/src/main/resources/rest-api-spec/test/cat.templates/10_basic.yaml
Comment (tlrx) [inline rest-api-spec/src/main/resources/rest-api-spec/test/cat.templates/10_basic.yaml]: Can you add more than 1 template?

**Review by tlrx — CHANGES_REQUESTED** [review]
@a2lin Thanks for your contribution! I left some minor comments but the overall looks good.

**Comment by a2lin** [comment]
@tlrx thanks for reviewing! I think I've made the requested changes.

**Review by tlrx — APPROVED** [review]

**Comment by tlrx** [comment]
@a2lin Thanks! Can you please squash the two commits into one? Once done I'll merge this PR and backport it to 5.0/5.x if @clintongormley is OK.

**Comment by tlrx** [comment]
@a2lin I squashed myself thanks to Github's famous merge button.

Thanks again for your contribution!

**Comment by tlrx** [comment]
Backported to 5.x in 144a7ca216d6da1859644b9c1544e931c9e9f5b4

**Pull request merged**
